### PR TITLE
Domain model

### PR DESCRIPTION
**Pull request by briandds**
This is a first cut at a domain model, described using an object model and tests.
Include with this PR:

* [Atat/Objects](https://github.com/dod-ccpo/skel/blob/913d9f9d19f0725471c48b6a833894f1128eb4ee/lib/Atat/Objects.pm) -- An object hierarchy for some of the resources in ATAT.
* [t/02-domain.t](https://github.com/dod-ccpo/skel/blob/913d9f9d19f0725471c48b6a833894f1128eb4ee/t/02-domain.t) -- A test using the above which includes aggregating eam expenses up to a task order.
* Changes to the UI which replace 'task order' with 'funding pool' (aka "bucket"), and roles which are now based on this generalization.

### CODE REVIEW

**Comment by briandds** [comment]
Review app: https://skel.atat.codes/domain_model

**Comment by jordangov** [comment]
We reviewed together, I think this is a good start, still lots to verify, but heading the right way.

File: lib/Atat/Objects.pm [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+# This file describes our domain using an object model.
+#
+# The following terminology may be helpful:
+#     * 'package' defines a class.
+#     * A class 'has' an attribute, which may be another object (delegation).
+#     * Plural attributes imply a 1-many (or many-many) relationship.
+#     * A class 'is' another class (vertical composition).
+#     * A class 'does' a role (horizontal composition).
+
+# User: An ATAT user.
```
Comment (hardikdds) [inline lib/Atat/Objects.pm:12]: I would like to edit the definition here to be:    

- [ ] - A user is uniquely identified by their email, and authenticated by their CAC card
- [ ] - A user can be any DoD employee or contractor

File: lib/Atat/Objects.pm [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  has 'roles';       # Roles within ATAT.
+}
+
+# UserRole: Defines a set of permissions surrounding access to a CSP.
+package UserRole {
+  use Mojo::Base -base;
+  has 'csp_account';    # CSP account (optional)
+  has 'permissions';    # List of permissions (strings) for this role.
+}
+
```
Comment (hardikdds) [inline lib/Atat/Objects.pm:26]: @briandds I would suggest that we...

- [ ] change the name of this object to "CSPUserRole" so as to not confuse it with "ATATUserRole"

File: lib/Atat/Objects.pm [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+# Account: An account.
+package Account {
+  use Mojo::Base -base;
+  has 'csp_role';       # The role in the CSP.
+  has 'token';          # Auth token.
+}
+
```
Comment (hardikdds) [inline lib/Atat/Objects.pm:33]: Changes to this: 

- [ ] This object should also be retitled to say "CSPAccount". 
- [ ] This object should have a Task Order number that it is referencing.  One Task Order number can be associated with multiple CSP accounts.

File: lib/Atat/Objects.pm [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+# Grant: A permission that a person has for a bucket.
+package Grant {
+  use Mojo::Base -base;
+  has 'user';           # An ATAT user.
+  has 'role';           # A UserRole (above).
+}
```
Comment (hardikdds) [inline lib/Atat/Objects.pm:39]: I would assume that a Grant needs a 'status' - such as Pending, Accepted, Rejected   and a 'history log' - Accepted on Jan 21st by user J.

File: lib/Atat/Objects.pm [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  has 'used';                # Amount of money used.
+  has 'grants';              # List of Grants for this bucket.
+  has 'distribute_into'
+    ;    # A list of other buckets into which this one can be distributed.
+  has '_aggregated_into'
+    ;    # A single parent bucket into which this one is aggregated.
+  has '_aggregated_from'    # The inverse of the above.
+    => sub { [] };
+  has 'csp_tags';           # Tags in the CSP associated with this bucket.
+
```
Comment (hardikdds) [inline lib/Atat/Objects.pm:58]: I'd say here that a 'bucket' should also have  'team tags'  - such as  "Navy"  "SPAWAR"  "Team Atlantic"

File: lib/Atat/Objects.pm (hunk 1 above)
Comment (jordangov) [reply lib/Atat/Objects.pm:12]: They will _not_ be uniquely identified by their email (although that should be a unique field). We will need a permanent identifier outside of email address as that can change.

File: lib/Atat/Objects.pm (hunk 4 above)
Comment (jordangov) [reply lib/Atat/Objects.pm:39]: Uuuh... I think we're getting ahead of ourselves on this one. We're going to have to talk a _lot_ more about approvals, acceptance, rejection, etc. This is just defining the relationship.

Comment (jordangov) [reply lib/Atat/Objects.pm:39]: (It's also a throw away model so we can get our feet wet, right? In which case, I don't care about logs, logs will come in time.)

File: lib/Atat/Objects.pm (hunk 2 above)
Comment (jordangov) [reply lib/Atat/Objects.pm:26]: Hmm... not sure I agree @hardikdds. We have two types of roles, yes, but we don't have a separate object for the two types, which might be just fine, but if we change this name then we need a new object as well for the other kind of role.

Comment (hardikdds) [reply lib/Atat/Objects.pm:26]: @jordangov    Thanks.  And Brian just explained to me how roles in most csps work.

**Comment by briandds** [comment]
@hardikdds and I discussed the issues above IRL, so I'm going to go ahead and merge this.

**Pull request merged**